src/http.c

```diff
@@ -553,102 +637,187 @@ static int http_header_select (char *h) {
 }
 
 #define PRINT_USERNAMES 1
+#define MAX_STRLEN 2048
 
-static void http_print_header(zfile f, char *header, unsigned length) {
-    char *token1, *token2, *token3, *saveptr;  
-    unsigned int not_first_header = 0;
-    enum http_type type = http_done;  
-    struct matches matches;
-
-    token1 = token2 = NULL; /* initialize just to avoid compiler warnings */
+static int http_parse_message(struct http_message *msg,
+                             char *data,
+                             unsigned int length) {
 
-    /* Start req/resp array */
-    zprintf(f, "[");
+    struct http_header *hdr = NULL;
+    char *token1, *token2, *token3, *saveptr = NULL;
+    enum http_type type = http_done;
+    unsigned int str_len = 0;
+    int i = 0;
 
     if (length < 4) {
-        /* End req/resp array */
-        zprintf(f, "]");
-        return;
+        return -1;
     }
 
+    if (msg == NULL) {
+        return -1;
+    }
 
     /*
-     * parse start-line, and print as request/status as appropriate
+     * Parse start-line, and get request/status lines.
      */
-    saveptr = header;
+    saveptr = data;
+
+    /* Easy access to the header storage */
+    hdr = &msg->header;
+
+    /* Try to get the start line of (header) data */
     type = http_get_start_line(&saveptr, &length, &token1, &token2, &token3);
-    if (type == http_request_line) {    
-    
-        zprintf(f, "{\"method\":\"%s\"},", token1);
+
+    if (type == http_malformed) {
+        return -1;
+    }
+
+    if (type == http_request_line) {
+        hdr->line_type = HTTP_LINE_REQUEST;
+
+        str_len = strnlen(token1, MAX_STRLEN);
+        hdr->line.request.method = calloc(str_len + 1, sizeof(char));
+        strncpy(hdr->line.request.method, token1, str_len);
+
+        str_len = strnlen(token2, MAX_STRLEN);
+        hdr->line.request.uri = calloc(str_len + 1, sizeof(char));
+        strncpy(hdr->line.request.uri, token2, str_len);
+
+        str_len = strnlen(token3, MAX_STRLEN);
+        hdr->line.request.version = calloc(str_len + 1, sizeof(char));
+        strncpy(hdr->line.request.version, token3, str_len);
+
+    } else if (type == http_status_line) {
+        hdr->line_type = HTTP_LINE_STATUS;
+
+        str_len = strnlen(token1, MAX_STRLEN);
+        hdr->line.status.version = calloc(str_len + 1, sizeof(char));
```

Comment:
Global comment check for NULL on all malloc/calloc calls

Reply:
agree